cmds/stty/termios.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2015-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"reflect"
+	"strconv"
+)
+
+type (
+	// tty is an os-independent version of the combined info in termios and window size structs.
+	// It is used to get/set info to the termios functions as well as marshal/unmarshal data
+	// in JSON formwt for dump and loading.
+	tty struct {
+		Ispeed int
+		Ospeed int
+		Row    int
+		Col    int
+		Line   int
+		Min    int
+		Time   int
+
+		CC map[string]uint8
+
+		Opts map[string]bool
+	}
+
+	winsize struct {
+		Row, Col       uint16
+		Xpixel, Ypixel uint16
+	}
+)
+
+func gtty(fd uintptr) (*tty, error) {
+	term, err := tiGet(fd)
+	if err != nil {
+		return nil, err
+	}
+	w, err := wsGet(fd)
+	if err != nil {
+		return nil, err
+	}
+
+	var t = tty{Opts: make(map[string]bool), CC: make(map[string]uint8)}
+	for n, b := range boolFields {
+		val := uint32(reflect.ValueOf(term).Elem().Field(b.word).Uint()) & b.mask
+		t.Opts[n] = val != 0
+	}
+
+	for n, c := range cc {
+		t.CC[n] = term.Cc[c]
+	}
+
+	// back in the day, you could have different i and o speeds.
+	// since about 1975, this has not been a thing. It's still in POSIX
+	// evidently. WTF?
+	t.Ispeed = int(term.Ispeed)
+	t.Ospeed = int(term.Ospeed)
+	t.Row = int(w.Row)
+	t.Col = int(w.Col)
+
+	return &t, nil
+}
+
+func stty(fd uintptr, t *tty) (*tty, error) {
+
+	// Get a syscall.Termios
+	// which we can partially fill in.
+	term, err := tiGet(fd)
+	if err != nil {
+		return nil, err
+	}
+
+	for n, b := range boolFields {
+		set := t.Opts[n]
+		i := reflect.ValueOf(term).Elem().Field(b.word).Uint()
+		if set {
+			i |= uint64(b.mask)
+		} else {
+			i &= ^uint64(b.mask)
+		}
+		reflect.ValueOf(term).Elem().Field(b.word).SetUint(i)
+	}
+
+	for n, c := range cc {
+		term.Cc[c] = t.CC[n]
+	}
+
+	term.Ispeed = uint32(t.Ispeed)
+	term.Ospeed = uint32(t.Ospeed)
+
+	if err := tiSet(0, term); err != nil {
```

Comment:
There's no guarantee that they are the same. Just because you asked the kernel to set something does not mean it sets it. 

If you feel that we should not do the gtty() at the end of stty() I am willing to be convinced but personally I find this behavior useful: you ask stty to set the fd and it returns what the kernel says really got set. Again, you can not assume the kernel did what you asked; the only way to know is to check.